File: applications/rtkdrawshepploganphantom/rtkdrawshepploganphantom.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtkdrawshepploganphantom_ggo.h"
#include "rtkGgoFunctions.h"

#include "rtkSheppLoganPhantomFilter.h"
#include "rtkDrawSheppLoganFilter.h"
#include "rtkConstantImageSource.h"
#include "rtkAdditiveGaussianNoiseImageFilter.h"

#include <itkImageFileReader.h>
#include <itkImageFileWriter.h>


int
main(int argc, char * argv[])
{
  GGO(rtkdrawshepploganphantom, args_info);

  constexpr unsigned int Dimension = 3;

  using OutputImageType = itk::Image<float, Dimension>;

  // Create a stack of empty projection images
  using ConstantImageSourceType = rtk::ConstantImageSource<OutputImageType>;
  auto constantImageSource = ConstantImageSourceType::New();
  rtk::SetConstantImageSourceFromGgo<ConstantImageSourceType, args_info_rtkdrawshepploganphantom>(constantImageSource,
                                                                                                  args_info);

  // Create a reference object (in this case a 3D phantom reference).
  using DSLType = rtk::DrawSheppLoganFilter<OutputImageType, OutputImageType>;
  DSLType::VectorType offset(0.);
  DSLType::VectorType scale;
  for (unsigned int i = 0; i < std::min(args_info.offset_given, Dimension); i++)
    offset[i] = args_info.offset_arg[i];
  scale.Fill(args_info.phantomscale_arg[0]);
  for (unsigned int i = 0; i < std::min(args_info.phantomscale_given, Dimension); i++)
    scale[i] = args_info.phantomscale_arg[i];
  auto dsl = DSLType::New();
  dsl->SetPhantomScale(scale);
  dsl->SetInput(constantImageSource->GetOutput());
  dsl->SetOriginOffset(offset);
  TRY_AND_EXIT_ON_ITK_EXCEPTION(dsl->Update())

  // Add noise
  OutputImageType::Pointer output = dsl->GetOutput();
  if (args_info.noise_given)
  {
    auto noisy = rtk::AdditiveGaussianNoiseImageFilter<OutputImageType>::New();
    noisy->SetInput(output);
    noisy->SetMean(0.0);
    noisy->SetStandardDeviation(args_info.noise_arg);
    TRY_AND_EXIT_ON_ITK_EXCEPTION(noisy->Update())
    output = noisy->GetOutput();
  }

  // Write
  TRY_AND_EXIT_ON_ITK_EXCEPTION(itk::WriteImage(output, args_info.output_arg))

  return EXIT_SUCCESS;
}

File: applications/rtkdrawshepploganphantom/rtkdrawshepploganphantom.py
#!/usr/bin/env python
import argparse
import sys
import itk
from itk import RTK as rtk


def build_parser():
    # Argument parsing
    parser = rtk.rtk_argument_parser("Computes a 3D voxelized Shepp & Logan phantom with noise [https://www.slaney.org/pct/pct-errata.html]")

    # General options
    parser.add_argument(
        "--verbose", "-v", help="Verbose execution", action="store_true"
    )
    parser.add_argument(
        "--output", "-o", help="Output projections file name", type=str, required=True
    )
    parser.add_argument(
        "--phantomscale",
        help="Scaling factor for the phantom dimensions",
        type=rtk.comma_separated_args(float),
        default=[128],
    )
    parser.add_argument("--noise", help="Gaussian noise parameter (SD)", type=float)
    parser.add_argument(
        "--offset",
        help="3D spatial offset of the phantom center",
        type=rtk.comma_separated_args(float),
    )

    rtk.add_rtk3Doutputimage_group(parser)

    # Parse the command line arguments
    return parser


def process(args_info: argparse.Namespace):
    # Define output pixel type and dimension
    OutputPixelType = itk.F
    Dimension = 3
    OutputImageType = itk.Image[OutputPixelType, Dimension]

    # Empty volume image
    constant_image_source = rtk.ConstantImageSource[OutputImageType].New()
    rtk.SetConstantImageSourceFromArgParse(constant_image_source, args_info)

    # Offset, scale, rotation
    offset = [0.0] * 3
    if args_info.offset:
        if len(args_info.offset) > 3:
            print("--offset needs up to 3 values", file=sys.stderr)
            sys.exit(1)
        offset = args_info.offset

    scale = [args_info.phantomscale[0]] * Dimension
    if args_info.phantomscale:
        if len(args_info.phantomscale) > 3:
            print("--phantomscale needs up to 3 values", file=sys.stderr)
            sys.exit(1)
        for i in range(len(args_info.phantomscale)):
            scale[i] = args_info.phantomscale[i]

    # Reference
    if args_info.verbose:
        print("Creating reference... ", flush=True)

    # DrawSheppLoganPhantomImageFilter
    dsl = rtk.DrawSheppLoganFilter[OutputImageType, OutputImageType].New()
    dsl.SetPhantomScale(scale)
    dsl.SetInput(constant_image_source.GetOutput())
    dsl.SetOriginOffset(offset)

    dsl.Update()

    # Add noise
    output = dsl.GetOutput()
    if args_info.noise:
        noisy = rtk.AdditiveGaussianNoiseImageFilter[OutputImageType].New()
        noisy.SetInput(output)
        noisy.SetMean(0.0)
        noisy.SetStandardDeviation(args_info.noise)

        noisy.Update()

        output = noisy.GetOutput()

    # Write
    if args_info.verbose:
        print("Writing reference... ", flush=True)

    itk.imwrite(output, args_info.output)


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    process(args_info)


if __name__ == "__main__":
    main()
